### PR TITLE
`version_compare()` operator can be null, return value depends on php version

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/4457

### CODE REVIEW

File: src/Type/Php/VersionCompareFunctionDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -4,22 +4,47 @@
...
 use PHPStan\Type\TypeUtils;
 use function array_filter;
 use function count;
+use function in_array;
 use function version_compare;
 
 class VersionCompareFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
 {
 
+	private $operators = [
```
Comment (staabm) [inline src/Type/Php/VersionCompareFunctionDynamicReturnTypeExtension.php:27]: all operators as defined in https://www.php.net/manual/en/function.version-compare.php

**Comment by staabm** [comment]
the positive cases are already covered in https://github.com/phpstan/phpstan-src/blob/fcfe9b9851fc640859aa02cebcfbfb84ee51624c/tests/PHPStan/Analyser/data/functions.php#L15-L22

**Comment by clxmstaab** [comment]
the remaining error is unrelated/fixed on master. should I rebase or do you just merge it as is?

File: tests/PHPStan/Analyser/data/bug-4457-php7.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	public function operatorCanBeNull() {
+		assertType('bool', version_compare('1.0', '1.1', null));
+	}
+
+	/**
+	 * @return int|bool
+	 */
+	function compare(string $version1, string $version2, ?string $comparator = null)
+	{
+		$result = version_compare($version1, $version2, $comparator);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/bug-4457-php7.php:21]: This regression test doesn't test anything here, you need to add assertType.

File: tests/PHPStan/Analyser/data/bug-4457.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+use function PHPStan\Testing\assertType;
+
+class Foo {
+	public function invalidOperator() {
+		assertType('*ERROR*', version_compare('1.0', '1.1', 'a'));
+	}
+
+	public function operatorCanBeNull() {
+		assertType('-1', version_compare('1.0', '1.1', null));
+	}
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/bug-4457.php:14]: Also test the #4457 scenario in this file.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: resources/functionMap_php80delta.php [hunk 4]
```diff
@@ -96,7 +96,7 @@
 		'strripos' => ['int|false', 'haystack'=>'string', 'needle'=>'string', 'offset='=>'int'],
 		'strrpos' => ['int|false', 'haystack'=>'string', 'needle'=>'string', 'offset='=>'int'],
 		'strstr' => ['string|false', 'haystack'=>'string', 'needle'=>'string', 'before_needle='=>'bool'],
-		'version_compare' => ['int|bool', 'version1'=>'string', 'version2'=>'string', 'operator='=>'string'],
+		'version_compare' => ['-1|0|1|bool', 'version1'=>'string', 'version2'=>'string', 'operator='=>'string'],
```
Comment (staabm) [inline resources/functionMap_php80delta.php:99]: I am not sure, should this also include `'operator'=>'string|null'`?

**Comment by staabm** [comment]
after making this return-type more precise we can see that a cross-php-version `version_compare()` check can get complicated, especially when phpstan-strict-types are activated.

any idea how this can be simplified?

**Comment by staabm** [comment]
Interessting case https://github.com/composer/composer/commit/dec123770ff1f5bde63f0a4b7ce1b5b627893bf8

**Pull request closed without merging**